### PR TITLE
Implement username/password authentication

### PR DESCRIPTION
**Pull request by chengyuhui**

### CODE REVIEW

**Comment by cssivision** [comment]
thanks for your pr, I plan to implement this feature, but i have no time recently. unfortunately, i can't get this pr merged directly, it's a big change. can you chop this into several prs? whatever thank you.

**Comment by chengyuhui** [comment]
To tell the truth, this is more like a partial rewrite, so it is not an easy job to split them ( none of the parts will work directly)
I will make comments on parts that I have changed.

File: Cargo.toml [hunk 1]
```diff
@@ -15,6 +15,10 @@ tokio-io = "0.1"
 tokio-timer = "0.1"
 
 [dev-dependencies]
-log = "0.3"
+log = "0.4"
 env_logger = "0.3"
 trust-dns-resolver = "0.7"
+
+[[example]]
```
Comment (chengyuhui) [inline Cargo.toml]: This makes it possible to do `cargo run --example socks5_proxy`

File: examples/socks5_proxy.rs [hunk 2]
```diff
@@ -25,62 +38,80 @@ fn main() {
     let mut lp = Core::new().unwrap();
     let handle = lp.handle();
     let listener = TcpListener::bind(&addr, &handle).unwrap();
-    let resolver = ResolverFuture::from_system_conf(&handle).unwrap();
+    let resolver = ResolverFuture::new(ResolverConfig::default(), ResolverOpts::default(), &handle);
```
Comment (chengyuhui) [inline examples/socks5_proxy.rs]: `ResolverFuture::from_system_conf` doesn't work on Windows, so I changed this.

File: examples/socks5_proxy.rs [hunk 3]
```diff
@@ -25,62 +38,80 @@ fn main() {
...
-    let streams = listener.incoming().and_then(|(socket, addr)| {
-        debug!("{}", addr);
-        tokio_socks5::serve(socket)
-    });
+    let streams = listener
+        .incoming()
+        .and_then(|(socket, _addr)| server.serve(socket));
+
+    let server = streams
+        .then(move |r| {
```
Comment (chengyuhui) [inline examples/socks5_proxy.rs]: The original version crashes if handshake fails, as `for_each` stops the stream when it hits an error.

File: examples/socks5_proxy.rs [hunk 4]
```diff
@@ -25,62 +38,80 @@ fn main() {
...
+            if r.is_err() {
+                println!("Error: {:?}", r);
+                return Ok(());
+            }
 
-    let server = streams.for_each(move |(c1, host, port)| {
-        println!("{}", addr);
-        println!("remote address: {}:{}", host, port);
+            if false {
```
Comment (chengyuhui) [inline examples/socks5_proxy.rs]: This may looks strange, it serves as an type hint for the compiler.

File: src/auth.rs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+use tokio_io::io::{read_exact, write_all};
+use std::io;
+
+use super::{check_version, other, read_var_len, FutureExt};
+use v5;
+
+pub trait AuthProvider {
+    /// Checks provided username & password.
+    fn check(&self, username: &str, password: &str) -> bool;
+    fn process<'a, C: AsyncRead + AsyncWrite + 'static>(
```
Comment (chengyuhui) [inline src/auth.rs]: This is written according to the RFC and tested against Telegram Desktop (since Chrome has no support for that).

File: src/server.rs [hunk 6]
```diff
@@ -1,137 +1,131 @@
...
-    pub const TYPE_IPV4: u8 = 1;
-    pub const TYPE_IPV6: u8 = 4;
-    pub const TYPE_DOMAIN: u8 = 3;
+use auth::{AuthProvider, NoopAuth};
+
+use super::{check_version, other, read_var_len, FutureExt};
+use v5;
+
+#[derive(Debug)]
+pub enum SocksAddr {
```
Comment (chengyuhui) [inline src/server.rs]: To make it suitable for others to use, I've decided not to return a string as address, which is hard to parse and hurts performance.

File: src/server.rs [hunk 7]
```diff
@@ -1,137 +1,131 @@
...
-    let timer = Timer::default();
-    let timeout = timer
-        .timeout(handshake_finish, Duration::new(10, 0))
-        .map_err(|_| other("handshake timeout"));
-
-    mybox(timeout.map(|(conn, addr, port)| (conn, addr, port)))
+/// A SOCKS5 server.
+pub struct Server<T: AuthProvider = NoopAuth> {
+    auth_provider: T,
+    handshake_timeout: u64,
```
Comment (chengyuhui) [inline src/server.rs]: Handshake timeout can be customized.

File: src/server.rs [hunk 8]
```diff
@@ -1,137 +1,131 @@
...
-use std::str;
-use std::net::{Ipv4Addr, Ipv6Addr};
 
 use futures::future;
 use futures::Future;
 use tokio_io::io::{read_exact, write_all};
-use tokio_core::net::TcpStream;
 use tokio_timer::Timer;
 
-mod v5 {
```
Comment (chengyuhui) [inline src/server.rs]: These are moved to `lib.rs`

File: src/server.rs [hunk 9]
```diff
@@ -1,137 +1,131 @@
...
+                } else {
+                    Err(other("unsupported command"))
+                }
+            })
+        });
+
+        // there's one byte which is reserved for future use, so we read it and discard it.
+        let resv = command.and_then(|c| read_exact(c, [0u8]));
+        let adress_type = resv.and_then(|(conn, _)| read_exact(conn, [0u8]));
+        let address = adress_type
```
Comment (chengyuhui) [inline src/server.rs]: ipaddr and host names are parsed using std methods
